driver/src/solution_submission/mod.rs

```diff
@@ -0,0 +1,132 @@
+use crate::contracts::stablex_contract::StableXContract;
+use crate::error::DriverError;
+
+use dfusion_core::models::{Order, Solution};
+
+use web3::types::U256;
+
+type Result<T> = std::result::Result<T, DriverError>;
+
+pub trait StableXSolutionSubmitting {
+    /// Return the objective value for the given solution in the given
+    /// batch or an error.
+    ///
+    /// # Arguments
+    /// - `batch_index` - the auction for which this solutions should be evaluated
+    /// - `orders` - the list of orders for which this solution is applicable
+    /// - `solution` - the solution to be evaluated
+    fn get_solution_objective_value(
+        &self,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
+    ) -> Result<U256>;
+
+    /// Submits the provided solution and returns the result of the submission
+    ///
+    /// # Arguments
+    /// - `batch_index` - the auction for which this solutions should be evaluated
+    /// - `orders` - the list of orders for which this solution is applicable
+    /// - `solution` - the solution to be evaluated
+    /// - `claimed_objective_value` - the objective value of the provided solution.
+    fn submit_solution(
+        &self,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
+        claimed_objective_value: U256,
+    ) -> Result<()>;
+}
+
+pub struct StableXSolutionSubmitter<'a> {
+    contract: &'a dyn StableXContract,
+}
+
+impl<'a> StableXSolutionSubmitter<'a> {
+    pub fn new(contract: &'a dyn StableXContract) -> Self {
+        Self { contract }
+    }
+}
+
+impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
+    fn get_solution_objective_value(
+        &self,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
+    ) -> Result<U256> {
+        self.contract
+            .get_solution_objective_value(batch_index, orders, solution)
```

Comment:
Maybe the `get_solution_objective_value` implementation should move here, and call the contract's `submit_solution` directly?

Reply:
Yes, for this PR I wanted to keep the factoring as is. Theoretically we could also minify the trait and just expose the submit solution method and have the fact that it's first verified and then sent hidden to the driver.

Reply:
Sounds like a great plan